### ISSUE TITLE
support bash style git completion for git command aliases containing a "null command"

### ISSUE THREAD

**Issue by rocketraman** [fish-shell/fish-shell#3954]
```
~ fish --version
fish, version 2.3.1
~ echo $version
2.3.1
```

Git completion for bash has a very useful feature in which a null command can be used to assign the completion command for an alias. Here is the description: https://github.com/git/git/blob/master/contrib/completion/git-completion.bash#L26.

It would be great if the git completion built in to fish would support this as well.

Note that, for some reason, I can't get this to work with bash, so I suppose its possible that my aliases are not set up correctly and fish already supports this, but I don't think that's the case.

**Comment by krader1961** [comment]
If I've correctly understood the comment you linked to this is done in fish using the `--wraps` flag. For example,

```
function f --wraps git
    do something
end
```

If that doesn't do what you want we're going to need more information about the behavior you're looking for.

**Comment by faho** [comment]
@krader1961: This is about a _git alias_ that calls a shell function. I.e. you'd define in the git config

    [alias]
        banana = "! somefunc () { : git commit; do something; }; somefunc"

which causes `git banana` to execute the defined "somefunc" function in a shell. I'm not sure how it determines that - it's possible it just uses $SHELL (which is usually not what you want).

Honestly, this doesn't sound supportable for us, because it requires parsing arbitrary (possibly POSIX-) shell script.

What could help here is to allow the `--wraps` option to wrap more than the command - in that case you could add `complete -c "git banana" -w "git commit"` and have completions working. That's #1976.

**Comment by krader1961** [comment]
Thanks, @faho, for pointing me in the right direction. Starting with your question about whether git honors $SHELL a simple experiment shows that git always uses /bin/sh:

```
$ git alias wtf '!testme() { echo hello; }; testme'
$ set -x GIT_TRACE 2
$ git wtf
17:55:11.120624 git.c:595               trace: exec: 'git-wtf'
17:55:11.121216 run-command.c:369       trace: run_command: 'git-wtf'
17:55:11.123133 run-command.c:369       trace: run_command: 'testme() { echo hello; }; testme'
17:55:11.124011 run-command.c:228       trace: exec: '/bin/sh' '-c' 'testme() { echo hello; }; testme' 'testme() { echo hello; }; testme'
hello
```

The bash git completion uses `git config --get alias.wtf` to get the definition of the alias and parses it to extract the completion type. See the `__git_aliased_command` function in the script pointed to in the original comment.

So we would need to parse both POSIX and fish syntax. Although the bash git completion script does require a very specific pattern; it doesn't actually parse arbitrary POSIX syntax. We could easily mimic that. It shouldn't be too hard to augment our `__fish_git_using_command` function to do what the bash git completion script is doing.

@rocketraman: Now that we've pointed you in the right direction do you want to take a stab at enhancing our git completion script?

**Comment by rocketraman** [comment]
> @rocketraman: Now that we've pointed you in the right direction do you want to take a stab at enhancing our git completion script?

I don't have much time right now to tackle this, but I'll keep it on my radar.

**Comment by rocketraman** [comment]
This is (mostly) fixed with #8129.
